File: Selection_Sort/Objective-C/SelectionSort.m
//
//  SelectionSort.m
//  Algorithms
//
//  Copyright © 2017 Grigory Lutkov <Friend.LGA@gmail.com>
//

#import "SelectionSort.h"

@implementation SelectionSort

+ (void)sort:(NSMutableArray<NSNumber *> *)array order:(SelectionSortOrder)order {
    NSUInteger arrayCount = array.count;

    if (arrayCount < 2) {
        return;
    }

    for (NSUInteger i = 0; i < arrayCount - 1; i++) {
        NSUInteger selectedIndex = i;
        NSNumber *selectedNumber = array[selectedIndex];

        for (NSUInteger j = i + 1; j < arrayCount; j++) {
            NSNumber *current = array[j];
            NSComparisonResult comparisonResult = [selectedNumber compare:current];

            if ((order == SelectionSortOrderAsc && comparisonResult == NSOrderedDescending) ||
                (order == SelectionSortOrderDesc && comparisonResult == NSOrderedAscending)) {
                selectedIndex = j;
                selectedNumber = array[selectedIndex];
            }
        }

        if (selectedIndex != i) {
            [array exchangeObjectAtIndex:i withObjectAtIndex:selectedIndex];
        }
    }
}

@end

File: Selection_Sort/Swift/SelectionSort.swift
//
//  SelectionSort.swift
//  Algorithms
//
//  Copyright © 2017 Grigory Lutkov <Friend.LGA@gmail.com>
//

import Foundation

struct SelectionSort {

    enum SortOrder {
        case ASC
        case DESC
    }

    /** Complexity is O(n^2) */
    static public func sort<T: Comparable>(_ array: [T?]?, order: SortOrder) -> [T?]? {
        let arrayCount = array?.count ?? 0

        if arrayCount < 2 {
            return array
        }

        var resultArray = array!

        for i in 0..<(arrayCount - 1) {
            var selectedIndex = i
            var selectedValue = resultArray[selectedIndex]

            for j in (i + 1)..<arrayCount {
                let currentValue = resultArray[j]

                if (order == .ASC && ((selectedValue != nil && currentValue == nil) || (selectedValue != nil && currentValue != nil && currentValue! < selectedValue!))) ||
                    (order == .DESC && ((selectedValue == nil && currentValue != nil) || (selectedValue != nil && currentValue != nil && currentValue! > selectedValue!))) {
                        selectedIndex = j
                        selectedValue = resultArray[selectedIndex]
                }
            }

            if (selectedIndex != i) {
                swap(&resultArray[i], &resultArray[selectedIndex])
            }
        }

        return resultArray;
    }

}
